File: frontend/src/services/resume.service.ts
import { api, handleApiError } from "./api";
import type {
  ResumeSummary,
  ResumeDetail,
  CreateResumeRequest,
  UpdateResumeRequest,
} from "../types/resume";
import type { ApiResponse } from "../types/api";

export const resumeService = {
  /**
   * Create a new resume
   */
  createResume: async (
    data: CreateResumeRequest
  ): Promise<ResumeDetail> => {
    try {
      const response = await api.post<ApiResponse<ResumeDetail>>(
        "/resumes",
        data
      );
      return response.data.data;
    } catch (error) {
      throw new Error(handleApiError(error));
    }
  },

  /**
   * Get all resumes for current user
   */
  getAllResumes: async (): Promise<ResumeSummary[]> => {
    try {
      const response = await api.get<ApiResponse<ResumeSummary[]>>("/resumes");
      return response.data.data;
    } catch (error) {
      throw new Error(handleApiError(error));
    }
  },

  /**
   * Get resume by ID
   */
  getResumeById: async (resumeId: number): Promise<ResumeDetail> => {
    try {
      const response = await api.get<ApiResponse<ResumeDetail>>(
        `/resumes/${resumeId}`
      );
      return response.data.data;
    } catch (error) {
      throw new Error(handleApiError(error));
    }
  },

  /**
   * Update resume
   */
  updateResume: async (
    resumeId: number,
    data: UpdateResumeRequest
  ): Promise<ResumeDetail> => {
    try {
      const response = await api.put<ApiResponse<ResumeDetail>>(
        `/resumes/${resumeId}`,
        data
      );
      return response.data.data;
    } catch (error) {
      throw new Error(handleApiError(error));
    }
  },

  /**
   * Delete resume
   */
  deleteResume: async (resumeId: number): Promise<void> => {
    try {
      await api.delete(`/resumes/${resumeId}`);
    } catch (error) {
      throw new Error(handleApiError(error));
    }
  },
};

File: Backend/src/main/java/com/resume/service/ResumeService.java
package com.resume.service;

import com.resume.dto.request.CreateResumeRequest;
import com.resume.dto.request.UpdateResumeRequest;
import com.resume.dto.response.ResumeResponse;
import com.resume.dto.response.ResumeSummaryResponse;
import com.resume.entity.Resume;
import com.resume.entity.User;
import com.resume.exception.ResumeNotFoundException;
import com.resume.exception.TemplateNotFoundException;
import com.resume.repository.ResumeRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ResumeService {

    private final ResumeRepository resumeRepository;
    private final AuthHelperService authHelperService;
    private final TemplateService templateService;

    @Transactional
    public ResumeResponse createResume(CreateResumeRequest request) {
        // Validate template exists
        if (!templateService.isValidTemplate(request.getTemplateId())) {
            throw new TemplateNotFoundException("Template not found with id: " + request.getTemplateId());
        }

        User currentUser = authHelperService.getCurrentLoggedInUser();

        Resume resume = new Resume();
        resume.setTitle(request.getTitle());
        resume.setTemplateId(request.getTemplateId());
        resume.setResumeDataJson(request.getResumeDataJson());
        resume.setUser(currentUser);

        Resume savedResume = resumeRepository.save(resume);

        return mapToResumeResponse(savedResume);
    }

    public List<ResumeSummaryResponse> getAllResumes() {
        User currentUser = authHelperService.getCurrentLoggedInUser();

        return resumeRepository.findAllByUserIdOrderByUpdatedAtDesc(currentUser.getId())
                .stream()
                .map(this::mapToResumeSummaryResponse)
                .collect(Collectors.toList());
    }

    public ResumeResponse getResumeById(Long resumeId) {
        User currentUser = authHelperService.getCurrentLoggedInUser();

        Resume resume = resumeRepository.findByIdAndUserId(resumeId, currentUser.getId())
                .orElseThrow(() -> new ResumeNotFoundException("Resume not found with id: " + resumeId));

        return mapToResumeResponse(resume);
    }

    @Transactional
    public ResumeResponse updateResume(Long resumeId, UpdateResumeRequest request) {
        // Validate template exists
        if (!templateService.isValidTemplate(request.getTemplateId())) {
            throw new TemplateNotFoundException("Template not found with id: " + request.getTemplateId());
        }

        User currentUser = authHelperService.getCurrentLoggedInUser();

        Resume resume = resumeRepository.findByIdAndUserId(resumeId, currentUser.getId())
                .orElseThrow(() -> new ResumeNotFoundException("Resume not found with id: " + resumeId));

        resume.setTitle(request.getTitle());
        resume.setTemplateId(request.getTemplateId());
        resume.setResumeDataJson(request.getResumeDataJson());

        Resume updatedResume = resumeRepository.save(resume);

        return mapToResumeResponse(updatedResume);
    }

    @Transactional
    public void deleteResume(Long resumeId) {
        User currentUser = authHelperService.getCurrentLoggedInUser();

        Resume resume = resumeRepository.findByIdAndUserId(resumeId, currentUser.getId())
                .orElseThrow(() -> new ResumeNotFoundException("Resume not found with id: " + resumeId));

        resumeRepository.delete(resume);
    }

    private ResumeResponse mapToResumeResponse(Resume resume) {
        return new ResumeResponse(
                resume.getId(),
                resume.getTitle(),
                resume.getTemplateId(),
                resume.getResumeDataJson(),
                resume.getCreatedAt(),
                resume.getUpdatedAt()
        );
    }

    private ResumeSummaryResponse mapToResumeSummaryResponse(Resume resume) {
        return new ResumeSummaryResponse(
                resume.getId(),
                resume.getTitle(),
                resume.getTemplateId(),
                resume.getUpdatedAt()
        );
    }
}